Configure logging using Logback's mechanism

georgberky opened the pull request:
Hi @raphw,

thanks for maintaining this project. We're consuming our logs using Datadog, a cloud based log aggregator and monitoring tool. It's much easier for those systems to consume log events when they're in machine readable format, so I've added an option to switch the log output to JSON hoping this might be handy for other people as well.

**Details:**
- activated using -J or --json-log
- uses Logstash's LoggingEventCompositeJsonEncoder
- required to up the Java Version to 11
- also updated Maven Plugins where possible
- also updates slf4j to 2.x

**More notes:**
I haven't been able to run all the tests on my machine (macOS) because of issues writing to the temporary directory.

In a spike I've managed to get one to work again using JUnit 5 and `@TempDir`. That also neatly got rid of some manual cleanup that the test was doing. I haven't added that to this PR because it would have conflated too many things.

However, I had to update Java to 11 in order to get the JSON log library to run (Logstash).

**Other considerations:**
I would love to have the option to let the user provide their own log configuration using either a `logback.xml` or even provide their own library at deploy time. The former seems like the next pragmatic thing to do. The latter is probably how they intended slf4j to be used.

Commit:
Add JSON logging

- activated using -J or --json-log
- uses Logstash's LoggingEventCompositeJsonEncoder
- required to up the Java Version to 11
- also updated Maven Plugins where possible

raphw commented:
Looks good, but could you remove the formatting changes that are not realted to the PR?

Commit:
Remove formatting changes

georgberky commented:
@raphw Hope I found all the formatting changes and reverted them. May I suggest the [Spotless Maven Plugin](https://github.com/diffplug/spotless/tree/main/plugin-maven) for the project? It formats the code according to the project's (or default) guidelines and makes formatting a non-issue. I've had good experiences with it. We're using something similar (ktlint) at work.

georgberky commented:
As for the logging, I think this prevents the standard mechanism from being used:

```
        //StandaloneServer.java:205 ff
        ch.qos.logback.classic.Logger rootLogger = loggerContext.getLogger(ch.qos.logback.classic.Logger.ROOT_LOGGER_NAME);
        loggerContext.stop();
        rootLogger.detachAndStopAllAppenders();
        rootLogger.addAppender(appender);
        rootLogger.setLevel(level);
        SLF4JBridgeHandler.removeHandlersForRootLogger();
        SLF4JBridgeHandler.install();
        LevelChangePropagator levelChangePropagator = new LevelChangePropagator();
        levelChangePropagator.setResetJUL(true);
        levelChangePropagator.setContext(loggerContext);
        levelChangePropagator.start();
        loggerContext.addListener(levelChangePropagator);
        loggerContext.start();
```

I like it as a fallback though so things just work out of the box. I have to read into the [logback documentation](https://logback.qos.ch/manual/configuration.html) a bit more, but this topic might warrant its own issue, efforts, and PRs.

Commit:
Fix formatting

georgberky commented:
I've played with `java -jar -Dlogback.configurationFile=…` when running the standalone server. It doesn't seem to work and uses the default configuration instead.

The way logback [looks for configuration](https://logback.qos.ch/manual/configuration.html#auto_configuration) is this:
- `Configurator` class via service locator
- `SerializedModelConfigurator`
  - `logback.scmoFile` property pointing to file
  - `logback-test.scmo` on the classpath
- `DefaultJoranConfigurator`
  - `logback.scmo` on the classpath
  - `logback.xml` on classpath
  - `logback-test.xml` on classpath
- `BasicConfigurator`
  - default logging to console

Maybe that's all we need already? What was the initial reason for configuring logback by hand?

Other idea: See if we can replace `BasicConfigurator` with our own default if the built-in ones are really bad.

raphw commented:
The current configuration was only put to make things work out of the box.

Now that I think of it: could you rather adjust the project to properly pick up an external configuration? Ideally, you'd make the same changes to the standalone client so they stay consistent.

Commit:
Standalone server: Use plain slf4j-logback logging

- also install jul-slf4j bridge at startup
- required for Glassfish/Grizzly server

Commit:
Remove logstash encoder dependency

Commit:
Remove hardcoded logging from standalone client

- configure logback using `java -Dlogback.configurationFile=/path/to/logback.xml -jar documents4j-client-standalone-<version>-shaded.jar <other options>`

georgberky commented:
Hi @raphw, I've updated the PR for both standalone client and server.

It now uses the default Logback mechanism for finding the configuration file. 

I've tested it with providing the path to the configuration file from the CLI as well:

```bash
java -Dlogback.configurationFile=/path/to/logback.xml -jar documents4j-client-standalone-<version>-shaded.jar <other options>
```

and

```bash
java -Dlogback.configurationFile=/path/to/logback.xml -jar documents4j-server-standalone-<version>-shaded.jar http://localhost:8080 -M
```

georgberky commented:
For end users who want to use something like JSON encoding or other encoders, this will mean that they'll have to add extra libraries to their deployment somehow. In case of the logstash encoder that would be logstash itself and transitively (IIRC) jackson for JSON encoding.

I've seen the [standalone server pubslished to Maven Central](https://central.sonatype.com/artifact/com.documents4j/documents4j-server-standalone). This is a good thing because people can now use tools like Groovy Grab or even curl to download it and all the dependencies required for their deployment.

All-in-all I think this is a good solution.

georgberky commented:
I didn't have to register the jul-slf4j bridge in the standalone client because it didn't seem to use any non-slf4j logging. Have I missed anything there?

raphw commented:
That might just be, I likely added it so it stays symmetric.

georgberky commented:
Hi @raphw , anything to add to this PR or to test?

raphw commented:
Could you undo the formatting changes as I suggested?

georgberky commented:
Woops. Thought I had found everything. 

I'll check again tomorrow. Thanks!

georgberky committed:
Fix formatting

georgberky commented:
@raphw Think I found the last ones. Could you check again?

raphw commented:
Looks good, one final change: where you replaced sys-out with log, there is still string concatenation used. Rather use {} and supply arguments.

georgberky commented:
Done. Hope I caught all of them.

georgberky committed:
Replace string concatenation with interpolation
